api/src/opentrons/protocol_engine/state/commands.py

```diff
@@ -88,96 +115,95 @@ def handle_action(self, action: Action) -> None:  # noqa: C901
             )
-            commands_by_id = self._state.commands_by_id.copy()
-            commands_by_id.update({queued_command.id: queued_command})
 
-            self._state = replace(self._state, commands_by_id=commands_by_id)
+            self._state.commands_by_id[queued_command.id] = queued_command
+            self._state.queued_command_ids[queued_command.id] = True
 
+        # TODO(mc, 2021-12-28): replace "UpdateCommandAction" with explicit
+        # state change actions (e.g. RunCommandAction, SucceedCommandAction)
+        # to make a command's queue transition logic easier to follow
         elif isinstance(action, UpdateCommandAction):
             command = action.command
-            commands_by_id = self._state.commands_by_id.copy()
-            commands_by_id.update({command.id: command})
 
-            self._state = replace(self._state, commands_by_id=commands_by_id)
+            self._state.commands_by_id[command.id] = command
+            self._state.queued_command_ids.pop(command.id, None)
+
+            if command.status == CommandStatus.RUNNING:
+                self._state.running_command_id = command.id
+            elif self._state.running_command_id == command.id:
+                self._state.running_command_id = None
 
         elif isinstance(action, FailCommandAction):
-            commands_by_id = self._state.commands_by_id.copy()
-            errors_by_id = dict(self._state.errors_by_id)
-            prev_command = commands_by_id[action.command_id]
-            command = prev_command.copy(
-                update={
-                    "errorId": action.error_id,
-                    "completedAt": action.failed_at,
-                    "status": CommandStatus.FAILED,
-                }
-            )
-            commands_by_id.update({command.id: command})
-            errors_by_id[action.error_id] = ErrorOccurrence.construct(
+            error_occurrence = ErrorOccurrence.construct(
                 id=action.error_id,
                 createdAt=action.failed_at,
                 errorType=type(action.error).__name__,
                 detail=str(action.error),
             )
 
-            self._state = replace(
-                self._state,
-                commands_by_id=commands_by_id,
-                errors_by_id=errors_by_id,
-            )
+            command_ids_to_fail = [
+                action.command_id,
+                *[i for i in self._state.queued_command_ids.keys()],
+            ]
+
+            for command_id in command_ids_to_fail:
+                prev_command = self._state.commands_by_id[command_id]
+                self._state.commands_by_id[command_id] = prev_command.copy(
+                    update={
+                        "errorId": action.error_id,
+                        "completedAt": action.failed_at,
+                        "status": CommandStatus.FAILED,
+                    }
+                )
+
+            if self._state.running_command_id == command_id:
```

Comment:
```suggestion
            if self._state.running_command_id == action.command_id:
```
Although, is this check appropriate? Shouldn't the `running_command_id` be cleared when an error occurs regardless of whether the error occurred while running this specific command? (Actually, when would `self._state.running_command_id == action.command_id` not be true?)

Reply:
Unmarking as resolved just to have the thread showing, but this was a good catch! Completely missed this bug.

> Although, is this check appropriate? Shouldn't the running_command_id be cleared when an error occurs regardless of whether the error occurred while running this specific command? (Actually, when would self._state.running_command_id == action.command_id not be true?)

This is a good question that I went back and forth on. It _would_ be an invalid state for a FailCommandAction to come in for a command that wasn't the `self._state.running_command_id`, so in normal operation this condition should always be true. I think this is a sign that this state structure will needs more tweaking to bake this requirement into the shape of the data itself